drivers: charger: Add TI BQ2407X driver

bjarki-andreasen opened the pull request:
Add driver for TI BQ2407X which is a standalone single cell charger with power path.

bjarki-andreasen committed:
drivers: charger: Add TI BQ2407X driver

Adds driver for TI BQ2407X

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,203 @@
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	switch (prop) {
+	case CHARGER_PROP_CHARGE_TYPE:
```
Is that documented somewhere? :)

Given the fact that you can seemingly set the status value to `CHARGER_STATUS_FULL` `CHARGER_STATUS_UNKNOWN` and that its named "status", I don't think this is intuitive?

rriveramcrus replied:
No, that's my bad. I can work on improving this documentation.

 This property is inherited from power_supply in Linux. From `linux/Documentation/ABI/testing/sysfs-class-power`:
```
What:		/sys/class/power_supply/<supply_name>/status
Date:		May 2007
Contact:	linux-pm@vger.kernel.org
Description:
		Represents the charging status of the battery. Normally this
		is read-only reporting although for some supplies this can be
		used to enable/disable charging to the battery.

		Access: Read, Write

		Valid values:
			      "Unknown", "Charging", "Discharging",
			      "Not charging", "Full"
```

I would like to maintain this common interpretation in Zephyr if that's alright with you.

bjarki-trackunit replied:
So making the driver accept setting "charging" and "not charging" to enable/disable charging then?

I usually agree with the way Linux does things, sysfs is not one of those things :) It is too flexible compared to dedicated APIs, leading to ambiguity and confusion as values within a single property themselves get properties (like r/w).

A dedicated charger_charge_enable(const struct device *dev, bool enabled); and having a charger_get_status(const struct device *dev, enum charger_status *status) with the "unknown" removed (the function can just return -ENOTSUP) would be unambiguous.

We are not limited by SYSFS like Linux is, everything is not a file on Zephyr, we can make actual functions that the linker can verify at build time :)

bbilas replied:
Then this would require refactoring the existing max20335-charger driver.
See: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/charger/charger_max20335.c

rriveramcrus replied:
>So making the driver accept setting "charging" and "not charging" to enable/disable charging then?

correct

>A dedicated charger_charge_enable(const struct device *dev, bool enabled); and having a charger_get_status(const struct device *dev, enum charger_status *status) with the "unknown" removed (the function can just return -ENOTSUP) would be unambiguous.

I'm not opposed to this, it has minimal impact for my future plans for this API. @aaronemassey, any concerns with this wrt what your team will be working on in the near future? This seems like one of those things that we could do better Linux.

>We are not limited by SYSFS like Linux is, everything is not a file on Zephyr, we can make actual functions that the linker can verify at build time :)

This is valid.

rriveramcrus replied:
sorry for the delay @bjarki-trackunit, I went ahead and drafted up a PR for the dedicated charge enable handler here #66128. Let me know what you think there.

rriveramcrus commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,203 @@
+		return -ENODEV;
+	}
+
+#ifdef CONFIG_BQ2407X_INIT_ENABLE_CHARGING
```
I feel like this is common of a feature that it could be a common charger dt property.

Would you mind writing a `charger.yaml` in `dts/bindings/charger/`?

bjarki-trackunit replied:
I will definitely add it :)

bjarki-trackunit replied:
Added

aasinclair commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,203 @@
+	 */
+	stat1 = gpio_pin_get_dt(&config->stat1_pin);
+	stat2 = gpio_pin_get_dt(&config->stat2_pin);
+	if ((stat1 && stat2) || stat1) {
```
```suggestion
	if (stat1) {
```

bjarki-trackunit replied:
Nice! applied

aasinclair commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,203 @@
+{
+	const struct ti_bq2407x_config *config = dev->config;
+
+	if (gpio_pin_get_dt(&config->pg_pin)) {
+		val->online = CHARGER_ONLINE_FIXED;
+	} else {
+		val->online = CHARGER_ONLINE_OFFLINE;
+	}
+}
```
gpio_pin_get_dt can fail, so you need to catch this and return it, e.g:

```suggestion
	int ret = gpio_pin_get_dt(&config->pg_pin);
	
	if (ret < 0) {
		return ret;
	}
	val->online = ret ? CHARGER_ONLINE_FIXED : CHARGER_ONLINE_OFFLINE;
	return 0;
}
```

Also applies to the other charge_prop functions.

bjarki-trackunit replied:
Added error handling of GPIO set/get

bjarki-andreasen committed:
dts: bindings: charger: Add enable-charging property

Charger drivers need to determine if charging shall be enabled
when initializing. This commit adds a base binding for chargers
which includes the boolean property "enable-charging" which will
make the driver enable charging by default when initializing.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

aasinclair commented on drivers/charger/Kconfig:
```diff
@@ -19,8 +19,13 @@ config CHARGER_INIT_PRIORITY
 	help
 	  Battery charger initialization priority.
 
+config CHARGER_INIT_ENABLE_CHARGING
```
This should be in devicetree rather than Kconfig

bjarki-trackunit replied:
Yeah I agree with that, will do

aasinclair reviewed (changes requested):
Looking good, just needs the charge init config moved to devicetree.

bbilas commented on drivers/charger/CMakeLists.txt:
```diff
@@ -8,3 +8,4 @@ zephyr_library_sources_ifdef(CONFIG_CHARGER_MAX20335 charger_max20335.c)
 zephyr_library_sources_ifdef(CONFIG_SBS_CHARGER sbs_charger.c)
 zephyr_library_sources_ifdef(CONFIG_USERSPACE charger_handlers.c)
 zephyr_library_sources_ifdef(CONFIG_EMUL_SBS_CHARGER emul_sbs_charger.c)
+zephyr_library_sources_ifdef(CONFIG_BQ2407X bq2407x.c)
```
Add a `CHARGER` prefix.

bjarki-trackunit replied:
Fixed

bbilas commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,216 @@
+		return -ENODEV;
+	}
+
+#ifdef CONFIG_CHARGER_INIT_ENABLE_CHARGING
```
It can be done using `if (IS_ENABLED(CONFIG_CHARGER_INIT_ENABLE_CHARGING)`

bbilas commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,216 @@
+
+#define TI_BQ2407X_DEVICE(inst)							\
+	struct ti_bq2407x_data ti_bq2407x_data##inst;				\
+	struct ti_bq2407x_config ti_bq2407x_config##inst = {			\
```
Missing `static const`.

```suggestion
	static const struct ti_bq2407x_config ti_bq2407x_config##inst = {			\
```

bbilas commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,216 @@
+}
+
+#define TI_BQ2407X_DEVICE(inst)							\
+	struct ti_bq2407x_data ti_bq2407x_data##inst;				\
```
Missing `static`

```suggestion
	static struct ti_bq2407x_data ti_bq2407x_data##inst;				\
```

bjarki-trackunit replied:
Fixed

bbilas commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,216 @@
+
+static int ti_bq2407x_init(const struct device *dev)
+{
+	struct ti_bq2407x_config *config = (struct ti_bq2407x_config *)dev->config;
```
```suggestion
	const struct ti_bq2407x_config *config = (struct ti_bq2407x_config *)dev->config;
```

bjarki-trackunit replied:
Fixed

bbilas commented on drivers/charger/CMakeLists.txt:
```diff
@@ -8,3 +8,4 @@ zephyr_library_sources_ifdef(CONFIG_CHARGER_MAX20335 charger_max20335.c)
 zephyr_library_sources_ifdef(CONFIG_SBS_CHARGER sbs_charger.c)
 zephyr_library_sources_ifdef(CONFIG_USERSPACE charger_handlers.c)
 zephyr_library_sources_ifdef(CONFIG_EMUL_SBS_CHARGER emul_sbs_charger.c)
+zephyr_library_sources_ifdef(CONFIG_CHARGER_BQ2407X bq2407x.c)
```
alphabetical order please (I know that CONFIG_ENUM is wrong but let's try to keep it sorted since now...).

bjarki-trackunit replied:
done

bbilas commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	if (config->enable_charging) {
+		ret = gpio_pin_configure_dt(&config->ce_pin, GPIO_OUTPUT_ACTIVE);
```
`ret` seems to be not needed. Just use `return gpio_pin_configure_dt(XXX`.

bjarki-trackunit replied:
ret removed

bbilas commented on drivers/charger/bq2407x.c:
```diff
@@ -0,0 +1,215 @@
+	const struct ti_bq2407x_config *config = dev->config;
+	int ret;
+
+	if ((gpio_pin_configure_dt(&config->stat1_pin, 0) < 0) ||
```
You should use `gpio_is_ready_dt` before that.

bjarki-trackunit replied:
added check for all gpios before pin configure

aaronemassey commented:
Just reviewed https://github.com/zephyrproject-rtos/zephyr/pull/66128, waiting for that to merge. Please feel free to "@" me. Apologies this got missed on my radar.

bjarki-trackunit commented:
### Update
* Rebased to include charge_enable API
* Replaced set_property API with charge_enable API (as this is the only set supported by the driver)
* Addressed @bbilas comments

aaronemassey reviewed (approved):
LGTM. @rriveramcrus or @GRobertZieba ?

rriveramcrus reviewed (approved):
nice work!

fabiobaltieri commented on dts/bindings/charger/ti,bq2407x.yaml:
```diff
@@ -0,0 +1,39 @@
+include: charger-device.yaml
+
+properties:
+  ce-gpios:
+    type: phandle-array
+    description: Charge enable GPIO
+    required: true
+
+  pg-gpios:
+    type: phandle-array
+    description: Power good GPIO
+    required: true
+
+  stat1-gpios:
+    type: phandle-array
+    description: Status 1 GPIO
+    required: true
+
+  stat2-gpios:
+    type: phandle-array
+    description: Status 2 GPIO
+    required: true
```
how about making these optional? maybe someone only wants power good or only charge status. or strap the CE to a rail...

fabiobaltieri commented on drivers/charger/charger_bq2407x.c:
```diff
@@ -0,0 +1,185 @@
+
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		ret = ti_bq2407x_get_prop_online(dev, val);
```
I think it'd be a lot cleaner if you resolve the union type here rather than in the function call:

`ret = ti_bq2407x_get_prop_online(dev, &val->online);`

that way is easier to review that the mapping is correct and the compiler takes care or the rest
